libs/utils/src/crashsafe.rs

```diff
@@ -79,12 +81,15 @@ pub fn path_with_suffix_extension(original_path: impl AsRef<Path>, suffix: &str)
         Some(extension) => Cow::Owned(format!("{extension}.{suffix}")),
         None => Cow::Borrowed(suffix),
     };
-    original_path
-        .as_ref()
-        .with_extension(new_extension.as_ref())
+    Utf8PathBuf::from_path_buf(
+        original_path
+            .as_ref()
+            .with_extension(new_extension.as_ref()),
+    )
+    .expect("non-Unicode path")
```

Comment:
To statically avoid this would require `original_path` to instead be `AsRef<Utf8Path>`. Have you explored what would that require? Looking at the pageserver changes at least, you've already changed all paths in PageserverConf, so I cannot really see where would these `std::path::Path` appear anymore for pageserver.

Seems the changes are not drastic:

```patch
diff --git a/libs/utils/src/crashsafe.rs b/libs/utils/src/crashsafe.rs
index 9a4c36f0b..dae6df90e 100644
--- a/libs/utils/src/crashsafe.rs
+++ b/libs/utils/src/crashsafe.rs
@@ -72,21 +72,15 @@ pub fn create_dir_all(path: impl AsRef<Utf8Path>) -> io::Result<()> {

 /// Adds a suffix to the file(directory) name, either appending the suffix to the end of its extension,
 /// or if there's no extension, creates one and puts a suffix there.
-pub fn path_with_suffix_extension(original_path: impl AsRef<Path>, suffix: &str) -> Utf8PathBuf {
-    let new_extension = match original_path
-        .as_ref()
-        .extension()
-        .map(OsStr::to_string_lossy)
-    {
+pub fn path_with_suffix_extension(
+    original_path: impl AsRef<Utf8Path>,
+    suffix: &str,
+) -> Utf8PathBuf {
+    let new_extension = match original_path.as_ref().extension() {
         Some(extension) => Cow::Owned(format!("{extension}.{suffix}")),
         None => Cow::Borrowed(suffix),
     };
-    Utf8PathBuf::from_path_buf(
-        original_path
-            .as_ref()
-            .with_extension(new_extension.as_ref()),
-    )
-    .expect("non-Unicode path")
+    original_path.as_ref().with_extension(&new_extension)
 }

 pub fn fsync_file_and_parent(file_path: &Utf8Path) -> io::Result<()> {
diff --git a/pageserver/src/tenant/mgr.rs b/pageserver/src/tenant/mgr.rs
index ef33afca1..d1c698206 100644
--- a/pageserver/src/tenant/mgr.rs
+++ b/pageserver/src/tenant/mgr.rs
@@ -70,12 +70,12 @@ impl TenantsMap {
 ///
 /// This is pageserver-specific, as it relies on future processes after a crash to check
 /// for TEMP_FILE_SUFFIX when loading things.
-async fn safe_remove_tenant_dir_all(path: impl AsRef<Path>) -> std::io::Result<()> {
+async fn safe_remove_tenant_dir_all(path: impl AsRef<Utf8Path>) -> std::io::Result<()> {
     let tmp_path = safe_rename_tenant_dir(path).await?;
     fs::remove_dir_all(tmp_path).await
 }

-async fn safe_rename_tenant_dir(path: impl AsRef<Path>) -> std::io::Result<Utf8PathBuf> {
+async fn safe_rename_tenant_dir(path: impl AsRef<Utf8Path>) -> std::io::Result<Utf8PathBuf> {
     let parent = path
         .as_ref()
         .parent()
@@ -92,7 +92,7 @@ async fn safe_rename_tenant_dir(path: impl AsRef<Path>) -> std::io::Result<Utf8P
         .collect::<String>()
         + TEMP_FILE_SUFFIX;
     let tmp_path = path_with_suffix_extension(&path, &rand_suffix);
-    fs::rename(&path, &tmp_path).await?;
+    fs::rename(path.as_ref(), &tmp_path).await?;
     fs::File::open(parent).await?.sync_all().await?;
     Ok(tmp_path)
 }
```